### DISCUSSION TITLE
Sharable/Public Forms

### DISCUSSION THREAD

**Discussion by benhaynes** [directus/directus#2977] [Feature Requests]
1. Add a share button to the Item Detail page (Create or Update) based on public permissions
    * Only shown when public permissions for the page/action are allowed
2. Creates a public URL including unguessable UUID token in an access parameter
    * If this feature were only for public permission data, then we might not need the UUID token to obfuscate the true resource URL... and we might be able to just use the normal App URL. However this is not extensible to private permission items/presets... and exposes potentially sensitive data.
3. That link can be sent to anyone, who can view/update the page

The page would be presented stripped down, with no module bar, no navigation bar, etc... just the page form (or preset listing in the future). This feature opens the door to "field work" where users collect data to a central server, sending polls to staff over email, and even embedding forms and datasets into public experiences (eg: website embed).

## Using a new system table
I'm always very cautious about adding a new system table, but this gives us all the flexibility we'd need.

```
directus_links
* id
* source_url
* token
* user_created
* date_created
* expires_on
* max_uses
* use_count
```

## Using JWTs
This is a nice/simple way to accomplish this feature, but doesn't allow for manually deleting the reference/token, doesn't allow for a "max uses" option, and doesn't track who created the share (or when) for accountability. These features aren't crucial for linking to public permission pages... but are required if/when we extend this to support private permission pages.

**Comment by Attacler** [comment]
This sounds good!
You could also use this for tickets or something.

I do have a question, how is it going to be displayed?

**Reply by Attacler** [reply]
Like it would be nice if you could customize it (maybe an extension?)

**Reply by benhaynes** [reply]
Yeah, basically a way to collect or display any data. I'm imagining the standard App design, but stripped down to remove the pieces that a public/scoped user wouldn't need to see (like modules/nav). I like the idea of having a more customizable/brandable header for these pages. 👍

**Comment by wendevlin** [comment]
This is a nice idea and would give the possibility to quickly collect public data.

I see so much possibilities, it's hard for me to write it down and think about everything.

I see concrete 2 features here:

**make collection or detail view "shareable"**

Displays the public user the collection or detail with a minimal user interface and no navigation.

- enables a share button on collection or detail view for users to send out view or edit permission to a public user.
- Should be done like on google docs
- Each collection should have an additional permission for this
- there should be an option for password protection
- I think the link can be always with a random unguessable UUID
- I am not sure if each user should send individual links, or are the links per item

**external Collection UI**

Give external users the chance to work/view in directus but limited to just one collection.
There should be the possibility to register/invite/login to a specific user group with specific permissions.
The navigation should be the minimal sidebar, so we can reuse the user settings for changing password or add other user information.

If the user is just allowed to add an item, then directus should automatically just show the create form. (this would be then useful for surveys). _The only problem I see here is that we need some kind of thank you page._

Options to set
- Optional detail view for user (collection only)
- Pre defined and lockable filters, sorts, .... (collection only)
- Optional information sidebar

**Reply by rijkvanzanten** [reply]
(Unrelated)

> I see so much possibilities, it's hard for me to write it down and think about everything.

This is the perfect summary of most of my days 😂

**Reply by rijkvanzanten** [reply]
> I see concrete 2 features here:

There's an option C: Offer a library / tool that allows you to inject a Directus form into any other app/website. Kinda similar to Netlify Forms

**Reply by benhaynes** [reply]
Thanks, @masterwendu, very insightful! As with most things in Directus, the key will be finding a simple, solid, and extensible MVP of this... since it can easily snowball into a huge thing that never gets started/completed. My thoughts on that:

### MVP
* Sharing Collection/Item Detail pages based on public role permissions only, uses standard page URL
* Info Sidebar component for sharing:
    * Shows page's public permissions (none, read, edit, create, etc)
    * Ability to copy sharable page link. For MVP, I imagine this is the normal URL, but appends a query string to identify it as a "share", so the user is not redirected to login. eg: `?share=public`
* Adjusted App view for these public pages (eg: remove/hide/adjust certain components that are not relevant)

### Added Over Time
* Ability to add/scope individual users
* Private URLs to pages (to mask PK, etc)
* Custom permissions per collection
* Password protection... which I think should be handled through a streamlined public registration flow (not a new/dedicated auth for sharing)
* Everything else...

I also like Rijk's idea to pull this out of Directus and be a standalone tool... we'll have to research what's ideal here.

**Reply by philleepflorence** [reply]
## Is this related to the Function of Google Forms and Spreadsheet?

A current work around (literally deployed yesterday) for a client that needs to _create_, _edit_, _publish_, _share_, and _embed_ public polls, surveys, RSVPs, and questionnaires.
The Google Spreadsheet(s) sync with Directus and act as "Middleware" to the DB Collections.

Gonna look also into [Netlify Forms](https://www.netlify.com/products/forms/) for ideas!

I can share a `Feature Mind Map`, `User Story`, and `Data Flow` if it helps...

**Reply by t7tran** [reply]
I'm so late to this party even though we have used [form.io](https://www.form.io/) for some time on the exact same line of thoughts.

While public forms are definitely useful in some use cases, I would be very wary. There are always jokers out there waiting to spam or ddos it.

According to my wild imagination, it would be something along this line:

- One can include global Vue into their application and create an application (unless it is already a Vue app).
- This application's mount serves as an entrypoint for the embedded form, i.e. collection form
- Existing Directus SDK can be used as a intermediator for authentication, proxy, etc. back to Directus backend

Ideally, if we can eliminate Vue dependency and wrap everything in a custom component, it opens doors. Or if Vue app/components can be packaged as a custom component.

**Reply by rijkvanzanten** [reply]
One interesting angle could be to allow `/fields` to be accessed in the form.io format, that way you could fetch a form.io configuration from Directus and include the form yourself however you see fit 🤔

**Comment by hermesalvesbr** [comment]
Integrating with something existing via directus cli would be a more practical option. use quasar or vuetify and a wizard that would choose public collections, and CRUD operations.

  1 - npx directus public create
2 - localhost:8055/public would be active
3 - Within directus you would configure public collections and permissions.

npx directus public create command would create a quasar or vuetify project already integrated into directus, available in /public and which could be modified, flexibly adjusted.

**Comment by abdonrd** [comment]
I see this feature in the [Directus Roadmap](https://directus.io/road-map/) from `Q4-2021`; but this functionality has not been released, right? Thanks!

**Reply by jamescammarano** [reply]
> I see this feature in the [Directus Roadmap](https://directus.io/road-map/) from `Q4-2021`; but this functionality has not been released, right? Thanks!

It has not been released.

**Reply by abdonrd** [reply]
Ok, thanks for the answer!

**Reply by rijkvanzanten** [reply]
> > I see this feature in the [Directus Roadmap](https://directus.io/road-map/) from `Q4-2021`; but this functionality has not been released, right? Thanks!
> 
> It has not been released.

That's not entirely true. We did releases _Shares_ since then, which allows you to share any item with somebody outside the platform. Shares are currently read-only though, which is why we don't consider this feature request fully done yet 🙂

**Reply by jamescammarano** [reply]
> > > I see this feature in the [Directus Roadmap](https://directus.io/road-map/) from `Q4-2021`; but this functionality has not been released, right? Thanks!
> > 
> > 
> > It has not been released.
> 
> That's not entirely true. We did releases _Shares_ since then, which allows you to share any item with somebody outside the platform. Shares are currently read-only though, which is why we don't consider this feature request fully done yet 🙂

Yeah, that's true. Progress has been made!

**Reply by abdonrd** [reply]
Thanks @rijkvanzanten! 👏🏻
Could you update the roadmap?

**Comment by NeoKactus** [comment]
This is more directed to Vuejs applications grabbing field data from Directus, but how feasible would it be to have Directus interface components available as a standalone package? Currently I have a prototype reading the field meta data from a Directus Collection and mapping the Directus field interfaces to Bootstrap-Vue using dynamic components. Currently works alright as a basic public facing form builder, but was wondering if there was a better way to go about this.

Basic idea:

`this.directusForm = await directus.fields.readMany('form_public_application_collection')`

```
<span v-for="field in directusForm">

      // field.meta.interface returns: select-dropdrown, select-radio, etc.
      // In my local nuxt application I have Bootstrap-Vue mapping components declared using the same naming scheme. 
      <component :is="field.meta.interface" v-model="form[field.meta.field]" :field="field" />
</span>
```

**Reply by rijkvanzanten** [reply]
>  but how feasible would it be to have Directus interface components available as a standalone package?

ohh interesting. Technically speaking, they're "just" vue components, though that being said, they're definitely purpose built to work well in the Directus app, and haven't really been designed with this out-of-context use. I think it might be quite a big undertaking to refactor those out enough to actually be useful outside of the Directus app 🤔

**Reply by NeoKactus** [reply]
Makes sense. I played around with sharing items, and it should suite my needs anyway once you can edit them. Thanks for the insight!

**Comment by sksar** [comment]
Any update on this one? Seems like the currently shared forms could also act on their own like a google form. 
Going with the "Netlify Forms" like feature is also pretty good, but it could lead to a lot of errors where there are like relationships and such.

**Reply by benhaynes** [reply]
We have sharable items (READ), but have not been focused on sharable forms (WRITE) yet. I agree that it would be huge though, it will likely need a PR or sponsor to get out in the coming months though, as we have a lot of other work going on at the moment.

**Comment by sksar** [comment]
@rijkvanzanten @benhaynes 
I believe that the publicly shareable forms should be a separate feature entirely, and should not be based on a collection.
In many cases where I needed forms, the exact form interface is vastly different from what is actually stored in the collections.
Also in many cases the forms that I would need would be, multiple forms for the same collection. Or even forms that would collect data across more than one collection.

So in my opinion forms should be a separate feature and should not be directly tied to the collection.
Like there could be a directus_forms table that would contain the form config, and also where to store the data. 

Also, forms should have an image uploading feature.

**Reply by LasseRosenow** [reply]
That sounds like a very good use case for the new `Flows` feature. Maybe it could get some sort of ability to specify public input values that can be shared and then within the flow you could specify what to do with all those input values 🤔

**Reply by benhaynes** [reply]
@sksar — ohh, interesting. That's a really great idea. I still think we'd want the ability to share the collection forms, but adding "alias forms" that can be shared and map to other collections/flows is a powerful concept. 👍

**Reply by sksar** [reply]
> @sksar — ohh, interesting. That's a really great idea. I still think we'd want the ability to share the collection forms, but adding "alias forms" that can be shared and map to other collections/flows is a powerful concept. 👍

Yes, basically it could be kinda like Typeform or something.

**Reply by benhaynes** [reply]
Not too crazy, either. We have all the parts: form builder, flows, public share pages, etc. But it's still non-trivial to stitch all this together.

**Reply by darioguarascio** [reply]
So basically a Directus version of typeform / jotform / Google forms.
You guys are smashing it. 

I'd love to have this.

**Comment by w0kyj** [comment]
Related Discussion: https://github.com/directus/directus/discussions/14949

**Comment by exsurgo** [comment]
I think the concept of **Plug-and-Play Components** could be a game-changer.    Being able to just plug in forms, data views, auth components, etc., into a front-end app and just have them work, potentially saves weeks/months of work.    And this would give Directus a huge competitive advantage over other BaaS's/Headless CMS's on the market.

I currently have an app that this would be perfect for.   I've [documented my ideas and use case here](https://docs.google.com/document/d/15QBrUPLtzyeUmAgEZ3m1sum6bXEXrXohu0NgcLNZrgs/edit?usp=sharing).

@benhaynes @rijkvanzanten  I'd be willing to prototype/assist on the web components and secure iframe for this.   I have a lot of experience in that area.

**Comment by Signpostproduct** [comment]
We would love if the shareable public forms had an interface for a map selection, because we have direct connectors to a mapping CMS in Directus. This would be much better than building out a custom interface on a google forms or Zendesk forms setup.

**Comment by jay-p-b-7span** [comment]
This is an exciting Feature. Would love to see this coming live 🎉
